contracts/tgrade-valset/src/contract.rs

```diff
@@ -677,14 +703,41 @@ fn end_block<Q: CustomQuery>(deps: DepsMut<Q>, env: Env) -> Result<Response, Con
 
     let old_validators = VALIDATORS.load(deps.storage)?;
 
-    VALIDATORS.save(deps.storage, &validators)?;
     // determine the diff to send back to tendermint
-    let (diff, add, remove) = calculate_diff(validators, old_validators);
+    let (diff, add, remove) = calculate_diff(validators.clone(), old_validators);
     let update_members = RewardsDistribution::UpdateMembers {
         add: add.clone(),
         remove: remove.clone(),
     };
 
+    if cfg.verify_validators {
+        // pending validators are validators who have just been added and have yet to verify they're online
+        // and signing blocks (or at least signing the first one)
+        let pending: Vec<_> = add
+            .iter()
+            .filter_map(|m| {
+                let addr = Addr::unchecked(&m.addr);
+                if let Ok(op) = operators().load(deps.storage, &addr) {
+                    if !op.active_validator {
+                        Some((addr, op.pubkey))
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                }
+            })
+            .collect();
+        PENDING_VALIDATORS.save(deps.storage, &pending)?;
```

Comment:
Oh, here you restore the set. So, no need to remove from it, above.

And, given that all this is being done only once per epoch, there are no races either. Nice.